File: controllers/recipeController.py
from fastapi import Response, HTTPException
from dotenv import load_dotenv
from starlette.status import HTTP_204_NO_CONTENT
from bson.objectid import ObjectId
import httpx
import os
import json

from models.recipe import Recipe
from models.userData import UserData
from schemas.recipe import recipeEntity, recipesEntity
import random

load_dotenv()
API_ID = os.getenv("EDAMAM_API_ID")
API_KEY = os.getenv("EDAMAM_API_KEY")
DBM_NAME = os.getenv("DBM_NAME")



def index(db, skip: int = 0, limit: int = 100):
    db = getattr(db, DBM_NAME)
    return recipesEntity(db.recipes.find())

def show(db, recipe_id: str):
    db = getattr(db, DBM_NAME)
    return recipeEntity(db.recipes.find_one({"_id": ObjectId(recipe_id)}))

def store(db, recipe: Recipe):
    db = getattr(db, DBM_NAME)
    new_recipe = dict(recipe)
    id = db.recipes.insert_one(new_recipe).inserted_id
    db_recipe = recipeEntity(db.recipes.find_one({"_id": ObjectId(id)}))
    return db_recipe

def update(db, recipe_id: str, recipe: Recipe):
    db = getattr(db, DBM_NAME)
    db_recipe = recipeEntity(db.recipes.find_one_and_update({"_id": ObjectId(recipe_id)}, {"$set": dict(recipe)}, return_document=True))

    return db_recipe

def delete(db, recipe_id: str):
    db = getattr(db, DBM_NAME)
    recipeEntity(db.recipes.find_one_and_delete({"_id": ObjectId(recipe_id)}))
    return Response(status_code=HTTP_204_NO_CONTENT)

async def getRecipeFromApi(userData: UserData):
    try:
        url = "https://api.edamam.com/api/recipes/v2"

        userData = dict(userData)
        params = {
            "type": "public",
            "beta": "false",
            "app_id": API_ID,
            "app_key": API_KEY
        }
        for key, value in userData.items():
            if value != None:
                if key == "nutrients":
                    for nut_key, nut_value in value.items():
                        if '%2B'in nut_value or '-' in nut_value:
                            params[f"{key}%5B{nut_key}%5D"] = f"{nut_value}"
                elif key == "query":
                    params["q"] = value
                else:
                    params[key] = value

        async with httpx.AsyncClient() as client:
            response = await client.get(url, params=params)

        if response.status_code == 200:
            data = dict(response.json())
            response = data.get("hits")
            
            if len(response) == 0:
                recipe = {"message": "No se encontraron recetas con los parámetros proporcionados"}
            else:
                response= random.choice(response)
                recipe = Recipe.from_dict(response["recipe"])
            

            return recipe
        else:
            raise HTTPException(status_code=response.status_code, detail=response.json())
    except httpx.HTTPError as err:
        raise HTTPException(status_code=500, detail=str(err))
    

File: app/Http/Controllers/Api/RecipeController.php
<?php

namespace App\Http\Controllers\Api;

use App\Enums\Health;
use App\Http\Controllers\Controller;
use App\Http\Requests\GetRecipeRequest;
use App\Http\Requests\StoreRecipeRequest;
use App\Http\Requests\UpdateRecipeRequest;
use App\Http\Resources\RecipeResource;
use App\Models\Patient;
use App\Models\Recipe;
use App\Models\User;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\Http;

class RecipeController extends Controller
{
    public function __construct()
    {
        $this->middleware(['can:recipe.view'])->only(['index', 'show']);
        //$this->middleware(['can:recipe.view_any'])->only('');
        $this->middleware(['can:recipe.create'])->only('store');
        $this->middleware(['can:recipe.update'])->only('update');
        $this->middleware(['can:recipe.delete'])->only('delete');
        $this->middleware(['can:recipe.generate'])->only('getRecipeFromApi');
    }
        /**
     * Muestra el listado de recetas
     * de los pacientes del nutricionista
     */
    public function index()
    {
        $patient = Patient::where('nutritionist_id', Auth::id())->pluck('patient_id');
        $recipes = Recipe::whereIn('user_id', $patient)
        ->orderBy('created_at','asc')
        ->paginate(15);

        return RecipeResource::collection($recipes);
    }

    /**
     * Store a newly created resource in storage.
     */
    public function store(StoreRecipeRequest $request)
    {
        $recipe = Recipe::create($request->validated());

        return new RecipeResource($recipe);
    }

    /**
     * Display the specified resource.
     */
    public function show(Recipe $recipe)
    {
        return new RecipeResource($recipe);
    }

    /**
     * Update the specified resource in storage.
     */
    public function update(UpdateRecipeRequest $request, Recipe $recipe)
    {
        $recipe->update($request->validated());

        return new RecipeResource($recipe);
    }

    /**
     * Remove the specified resource from storage.
     */
    public function destroy(Recipe $recipe)
    {
        $recipe->delete();

        return response()->json([
            'message' => 'Receta eliminada satisfactoriamente',
        ]);
    }

    public function getRecipeFromApi(GetRecipeRequest $request)
    {
        try {
            $auth_user = User::find($request->input('user_id'));
            if ($auth_user->hasRole('paciente')) {
                $nutritional_profile = $auth_user->nutritionalProfile;

                $request['health'] = $nutritional_profile->allergies;
            }


            $params = [
                'type' => 'public',
                'beta' => false,
                'app_id' => config('recipe_api.edamam.id'),
                'app_key' => config('recipe_api.edamam.key'),
            ];

            $fields = [
                "label",
                "dietLabels",
                "healthLabels",
                "dishType",
                "cautions",
                "ingredientLines",
                "calories",
                "totalTime",
            ];

            // Añadir los parámetros adicionales del usuario
            foreach ($request->all() as $key => $value) {
                if ($value !== null) {
                    if ($key === "nutrients") {
                        foreach ($value as $nut_key => $nut_value) {
                            if (strpos($nut_value, '%2B') !== false || strpos($nut_value, '-') !== false) {
                                $params[$key . "%5B" . $nut_key . "%5D"] = $nut_value;
                            }
                        }
                    } elseif ($key === "query") {
                        $params["q"] = $value;
                    } else {
                        $params[$key] = $value;
                    }
                }
            }
            $url = "https://api.edamam.com/api/recipes/v2?" . http_build_query($params);

            foreach ($fields as $field) {
                $url .= '&field=' . urlencode($field);
            }
            // Realizar la solicitud a la API
            $response = Http::get($url);
            if ($response->successful()) {
                $data = $response->json();
                $hits = $data['hits'] ?? [];
                if (count($hits) === 0) {
                    $recipe = response()->json([
                        "message" => "No se encontraron recetas con los parámetros proporcionados"
                    ]);
                } else {
                    $recipe = $hits[array_rand($hits)];
                    if($request->filled('patient_id')){
                        $patient = User::find($request->input('patient_id'));
                        if ($patient->hasRole('paciente')){
                            $recipe['user_id'] = $request->input('patient_id');
                        }
                    }
                }

                return response()->json($recipe);
            } else {
                return response()->json($response->json(), $response->status());
            }
        } catch (\Exception $e) {
            return response()->json([
                'message' => 'Error al obtener recetas de la API',
                'error' => $e->getMessage(),
            ], 500);
        }
    }
}
